### ISSUE TITLE
Warning -- asyncore.socket_map was modified by test_ssl

### ISSUE THREAD

**Issue by vstinner** [python/cpython#74385]
BPO | [30199](https://bugs.python.org/issue30199)
--- | :---
Nosy | @vstinner
PRs | <li>python/cpython#1381</li><li>python/cpython#1389</li><li>python/cpython#1390</li><li>python/cpython#1408</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-05-03.10:34:25.673>
created_at = <Date 2017-04-28.13:38:20.838>
labels = ['3.7', 'tests', 'performance']
title = 'Warning -- asyncore.socket_map was modified by test_ssl'
updated_at = <Date 2017-05-03.10:34:25.671>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-05-03.10:34:25.671>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-05-03.10:34:25.673>
closer = 'vstinner'
components = ['Tests']
creation = <Date 2017-04-28.13:38:20.838>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 30199
keywords = []
message_count = 6.0
messages = ['292532', '292746', '292770', '292771', '292830', '292883']
nosy_count = 1.0
nosy_names = ['vstinner']
pr_nums = ['1381', '1389', '1390', '1408']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'resource usage'
url = 'https://bugs.python.org/issue30199'
versions = ['Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
It seems like test_asyncore_server() of test_ssl doesn't cleanup properly asyncore on FreeBSD, and so following unit tests can be impacted (and fail).

http://buildbot.python.org/all/builders/AMD64%20FreeBSD%20CURRENT%20Debug%203.x/builds/210/steps/test/logs/stdio

test_asyncore_server (test.test_ssl.ThreadedTests)
Check the example asyncore integration. ... 
 server:  new connection from 127.0.0.1:48985
 client:  sending b'FOO\\n'...
 server:  read b'FOO\\n' from client
 client:  read b'foo\\n'
 client:  closing connection.
 server:  read b'over\\n' from client
 client:  connection closed.
 cleanup: stopping server.
 cleanup: joining server thread.
 cleanup: successfully joined.
ok
...
Warning -- asyncore.socket_map was modified by test_ssl
  Before: {}
  After:  {6: \<test.test_ssl.AsyncoreEchoServer.EchoServer.ConnectionHandler 127.0.0.1:48985 at 0x80664c468\>}

Maybe AsyncoreEchoServer.\_\_exit__() should just ends with "asyncore.close_all(ignore_all=True)"?

**Comment by vstinner** [comment]
New changeset 1dae7450c68bad498e57800387b24cb103c461fa by Victor Stinner in branch 'master':
bpo-30199: test_ssl closes all asyncore channels (bpo-1381)
https://github.com/python/cpython/commit/1dae7450c68bad498e57800387b24cb103c461fa

**Comment by vstinner** [comment]
New changeset 89a54c73746417bae003a9985668fa665040d5d9 by Victor Stinner in branch '3.5':
bpo-30199: test_ssl closes all asyncore channels (bpo-1381) (bpo-1389)
https://github.com/python/cpython/commit/89a54c73746417bae003a9985668fa665040d5d9

**Comment by vstinner** [comment]
New changeset 3c422b92653df05c7034136b77f76376aa5aec4f by Victor Stinner in branch '3.6':
bpo-30199: test_ssl closes all asyncore channels (bpo-1381) (bpo-1390)
https://github.com/python/cpython/commit/3c422b92653df05c7034136b77f76376aa5aec4f

**Comment by vstinner** [comment]
New changeset d1c862ffa73b222746035155817d2516c4f5b71e by Victor Stinner in branch '2.7':
bpo-30199: test_ssl closes all asyncore channels (bpo-1381) (bpo-1408)
https://github.com/python/cpython/commit/d1c862ffa73b222746035155817d2516c4f5b71e

**Comment by vstinner** [comment]
Fix applied to 2.7, 3.5, 3.6 and master (3.7).
